README.md
# Ping Bong

[![npm](https://img.shields.io/npm/l/ping-bong.svg)](https://www.npmjs.org/package/ping-bong) [![npm](https://img.shields.io/npm/v/ping-bong.svg)](https://www.npmjs.org/package/ping-bong)
![Node version](https://img.shields.io/node/v/ping-bong.svg?style=flat-square)

A simple ping tool that follows HTTP redirections.

## 🏓 Installation

```shell
$ npm install -g ping-bong
```

## 🏓 Usage

## CLI

```shell
$ ping-bong http://www.softonic.com/ie/12345
[
  {
    "method": "head",
    "url": "http://www.softonic.com/ie/12345",
    "to": "https://www.softonic.com/ie/12345",
    "statusCode": 301,
    "statusText": "Moved Permanently",
    "responseTime": 126,
    "userAgent": "ping-bong/3.0.5"
  },
  {
    "method": "head",
    "url": "https://www.softonic.com/ie/12345",
    "to": "https://pingtool.softonic.com",
    "statusCode": 301,
    "statusText": "Moved Permanently",
    "responseTime": 313,
    "userAgent": "ping-bong/3.0.5"
  },
  {
    "method": "head",
    "url": "https://pingtool.softonic.com",
    "statusCode": 200,
    "statusText": "OK",
    "responseTime": 300,
    "userAgent": "ping-bong/3.0.5"
  }
]

```

## API

```javascript
const PingBong = require('ping-bong');

(async () => {
  const pingyBong = new PingBong({
    httpOptions = {
      method: 'get',
      headers: {
        'User-Agent': 'pingee-beengee/42.0',
      },
      maxRedirects: 10,
    },
    includes = {
      'request.method': 'method',
      'request.url': 'url',
      'response.headers.location': 'to',
      'response.status': 'statusCode',
      'response.statusText': 'statusText',
      'response.responseTime': 'responseTime',
      'request.headers.user-agent': 'userAgent',
    },
  });

  const redirections = await pingyBong.check({ url });

  console.log(JSON.stringify(redirections, null, 2));
})();
```

## 🏓 Demo

Clone the project...

```shell
$ git clone https://github.com/mawrkus/ping-bong.git
$ cd bong
$ npm install
$ npm run demo
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [3.0.5] - 2018-12-07

### Changed

- Updated dependencies

## [3.0.2] -> [3.0.4] - 2018-xx-xx

- ...

## [3.0.1] - 2018-05-04

### Added

- Default "error" statusText on error responses

## [3.0.0] - 2018-05-03

### Changed

- The `includes` parameter is now a mapping between 2 lodash paths to get values from the request/response and set them to the output result

## [2.0.0] - 2018-05-03

### Added

- `htttpOptions` and `includes` options

### Changed

- Switched implementation to a class
- Moved `method`, `url` and `userAgent` options to `httpOptions`
- Reformat output to be a single array which items have the same format

## [1.0.0] - 2018-03-21

### Added

- Initial release as a simple "bong" function


bong.js
#!/usr/bin/env node

const axios = require('axios');
const debug = require('debug')('bong');
const { name, version } = require('./package.json');

const bongAgent = `${name} v${version}`

module.exports = async function bong({ url, userAgent = bongAgent, redirections = [] } = {}) {
  try {
    const requestOptions = {
      method: 'head',
      url,
      maxRedirects: 0,
      headers: {
        'User-Agent': userAgent,
      }
    };

    const { status: statusCode, statusText } = await axios.request(requestOptions);

    debug('HEAD %s: %s (%d)', url, statusText, statusCode);

    redirections.push({ url, statusCode });

    return redirections;
  } catch (error) {
    const { status: statusCode, statusText, headers } = error.response;

    debug('HEAD %s: %s (%d)', url, statusText, statusCode);

    if (statusCode >= 300 && statusCode < 400) {
      const { location: to } = headers;

      redirections.push({ to, statusCode });

      debug('Following "%s"...', to);

      return bong({ url: to, userAgent, redirections });
    }

    redirections.push({ error: { statusCode, statusText }, url });

    return redirections;
  }
}


index.js
#!/usr/bin/env node

const axios = require('axios');
const debug = require('debug')('ping');

async function ping({ url, redirections = [] } = {}) {
  try {
    const { status: statusCode, statusText } = await axios.request({
      method: 'head',
      url,
      maxRedirects: 0,
    });

    debug('HEAD %s: %s (%d)', url, statusText, statusCode);

    redirections.push({ statusCode, url });

    return redirections;
  } catch (error) {
    const { status: statusCode, statusText, headers } = error.response;
    const { location: to } = headers;

    debug('HEAD %s: %s (%d)', url, statusText, statusCode);

    if (statusCode >= 300 && statusCode < 400) {
      redirections.push({ statusCode, to });
      debug('Following "%s"...', to);
      return ping({ url: to, redirections });
    }

    redirections.push({ error: { statusCode, statusText }, url });

    return redirections;
  }
}

(async () => {
  const url = process.argv[2];
  const redirections = await ping({ url });
  console.log(JSON.stringify({ url, redirections }, null, 2));
})();


es/ping-bong.js
#!/usr/bin/env node

const axios = require('axios');
const get = require('lodash.get');
const set = require('lodash.set');
const debug = require('debug')('ping-bong');
const { name: pkgName, version: pkgVersion } = require('../package.json');

/**
 * @param  {AxiosResponse} response
 * @return {boolean}
 */
function isRedirection(response) {
  const { status } = response;
  return (status >= 300 && status < 400);
}

/**
 * @param {AxiosResponse} response
 */
function logResponse({ config = {}, status = '?', statusText = '?' /* , headers */ }) {
  const { method = '?', url = '?' } = config;
  debug('%s %s: %s (%d)', method.toUpperCase(), url, statusText, status);
  // debug(headers);
}

/**
 * @param {AxiosRequest} request
 */
function logRequest({ method, url/* , headers */ }) {
  debug('%s %s...', method.toUpperCase(), url);
  // debug(headers);
}

/**
 * @param  {Object} httpOptions
 * @return {HttpClient}
 */
function createHttpClient(httpOptions) {
  const httpClient = axios.create({
    ...httpOptions,
    maxRedirects: 0, // no follow
  });

  httpClient.interceptors.request.use((config) => {
    logRequest(config);
    return config;
  }, async (error) => {
    debug('Request error!');
    debug(error);
    throw error;
  });

  httpClient.interceptors.response.use((response) => {
    logResponse(response);
    return response;
  }, async (error) => {
    const { response = {} } = error;
    logResponse(response);
    throw error;
  });

  // debug('HTTP client created');
  // debug(httpClient.defaults);

  return httpClient;
}

module.exports = class PingBong {
  /**
   * @param {Object} [httpOptions]
   * @param {String} [httpOptions.method='head']
   * @param {Object} [httpOptions.headers]
   * @param {string} [httpOptions.headers.User-Agent]
   * @param {Object} [includes]
   */
  constructor({
    httpOptions = {
      method: 'head',
      headers: {
        'user-agent': `${pkgName}/${pkgVersion}`,
      },
      maxRedirects: 10,
    },
    includes = {
      'request.method': 'method',
      'request.url': 'url',
      'response.headers.location': 'to',
      'response.status': 'statusCode',
      'response.statusText': 'statusText',
      'response.responseTime': 'responseTime',
      'request.headers.user-agent': 'userAgent',
    },
  } = {}) {
    debug('🏓  %s v%s', pkgName, pkgVersion);

    const { maxRedirects } = httpOptions;
    this._maxRedirects = Number(maxRedirects) >= 0 ? Number(maxRedirects) : 10;

    this._httpOptions = { ...httpOptions };
    delete this._httpOptions.maxRedirects;

    this._httpClient = createHttpClient(httpOptions);

    this._includes = { ...includes };

    debug('HTTP options', this._httpOptions);
    debug('Max redirects =', this._maxRedirects);
    debug('Includes', this._includes);
  }

  /**
   * @param  {string}  url
   * @return {Promise.<Object[]>}
   */
  async check({ url }) {
    const redirections = [];
    let currentUrl = url;
    let remainingRedirects = this._maxRedirects;

    /* eslint-disable no-await-in-loop */
    do {
      const start = Date.now();

      try {
        const response = await this._httpClient.request({ url: currentUrl });

        response.responseTime = Date.now() - start;

        const finalRedirection = this._grabRedirection({ response });

        redirections.push(finalRedirection);

        currentUrl = null;
      } catch (error) {
        const { response = { statusText: 'error' } } = error;

        response.responseTime = Date.now() - start;

        const redirection = this._grabRedirection({ response });

        if (isRedirection(response)) {
          debug('Max %d redirect(s) to follow.', remainingRedirects);
          remainingRedirects -= 1;
          currentUrl = response.headers.location;
        } else {
          redirection.url = currentUrl;
          redirection.error = error.message;
          currentUrl = null;
        }

        redirections.push(redirection);
      }
    } while (currentUrl && remainingRedirects >= 0);
    /* eslint-enable no-await-in-loop */

    return redirections;
  }

  /**
   * @param  {AxiosResponse} response
   * @return {Object}
   */
  _grabRedirection({ response }) {
    const syntheticResponse = {
      request: response.config,
      response,
    };
    const redirection = {};

    Object.entries(this._includes)
      .forEach(([from, to]) => {
        const value = get(syntheticResponse, from);
        set(redirection, to, value);
      });

    return redirection;
  }
};
